library/std/src/net/addr.rs

```diff
@@ -290,7 +290,11 @@ impl SocketAddrV4 {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn ip(&self) -> &Ipv4Addr {
-        unsafe { &*(&self.inner.sin_addr as *const c::in_addr as *const Ipv4Addr) }
+        // SAFETY:
+        //   * `in_addr` is `struct { _: [u8; 4]; }` in most OSes.
+        //   * `Ipv4Addr` is `struct { _: u32 }`.
+        //   * Both are `#[repr(C)]` and have the same size and alignment.
+        unsafe { &*((&self.inner.sin_addr as *const c::in_addr).cast::<Ipv4Addr>()) }
```

Comment:
I added a few unsafety explanation.
But is it correct?

Reply:
I don't quite understand how it works.